src/gil.rs

```diff
@@ -10,15 +10,29 @@ use std::{mem, ptr::NonNull, sync::atomic};
 
 static START: Once = Once::new();
 
-thread_local! {
+cfg_if::cfg_if! {
+    if #[cfg(thread_local_const_init)] {
+        use std::thread_local as thread_local_const_init;
+    } else {
+        macro_rules! thread_local_const_init {
+            ($(#[$attr:meta])* static $name:ident: $ty:ty = const { $init:expr };) => (
+                thread_local! { $(#[$attr])* static $name: $ty = $init; }
+            )
+        }
+    }
+}
+
+thread_local_const_init! {
     /// This is an internal counter in pyo3 monitoring whether this thread has the GIL.
     ///
     /// It will be incremented whenever a GILGuard or GILPool is created, and decremented whenever
     /// they are dropped.
     ///
     /// As a result, if this thread has the GIL, GIL_COUNT is greater than zero.
-    static GIL_COUNT: Cell<usize> = Cell::new(0);
+    static GIL_COUNT: Cell<usize> = const { Cell::new(0) };
+}
 
+thread_local! {
     /// Temporarily hold objects that will be released when the GILPool drops.
     static OWNED_OBJECTS: RefCell<Vec<NonNull<ffi::PyObject>>> = RefCell::new(Vec::with_capacity(256));
```

Comment:
I wonder whether we should apply this here as well, of course using `RefCell::new(Vec::new())` as the initializer. The flag needs to be checked every time while the initial allocation should be amortized after the first few calls.

Reply:
(Maybe to add: I am not a huge fan of unconditionally pre-allocating long-lived buffers in the first place. I think pre-allocation is best applied when I have something to base the size in and am sure that I will use the allocation, e.g. collecting an iterator and pre-allocating based on its size hint.)

Reply:
It could be worth a go to see if this does affect benchmarks in any way, I think we likely access this more frequently than `GIL_COUNT`.

I agree with you about the long-lived buffers; I am very much hoping to resurrect work on #1308 come mid-summer with a goal of us coming up with the adjusted object API which doesn't use this mechanism.

Reply:
I've made `OWNED_OBJECTS` to be const-initialized as well in 291fc36d0c6a62acfceb58255b3d5da11e1dfb55.